Move some bmm/baddbmm to ATen

t-vi opened the pull request:
- Incorporates MKL addition by @mingfeima  Thank you! (but all errors are my own)
- Native CPU implementation: defer to matrix multiplication for
  small batches and parallelize over batch dimension for large
  batches.
- Add bmm test for CUDA just to be sure.

This is a partial fix for #10661, getting down to a factor ~5.
Considerable overhead is incurred for the setup in einsum. It might
be more efficient to eventually define an optimized contraction
functions for arbitrary and several dimensions.

t-vi committed:
Move some bmm/baddbmm to ATen

- Native CPU implementation: defer to matrix multiplication for
  small batches and parallelize over batch dimension for large
  batches.
  (More improvement might be achieved by calling MKL's batch gemm.)
- Add bmm test for CUDA just to be sure.

This is a partial fix for #10661, getting down to a factor ~5.
Considerable overhead is incurred for the setup in einsum. It might
be more efficient to eventually define an optimized contraction
functions for arbitrary and several dimensions.

t-vi committed:
Merge branch 'master' into bmm_manysmall

vishwakftw commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -222,6 +228,145 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+  auto s0 = self.accessor<scalar_t, 3>();
+  auto m0 = mat2.accessor<scalar_t, 3>();
+
+  #pragma omp parallel for if(bs > 100) // or small ks?
```
I think this needs to be guarded.

vishwakftw commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -222,6 +228,145 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+  auto s0 = self.accessor<scalar_t, 3>();
+  auto m0 = mat2.accessor<scalar_t, 3>();
+
+  #pragma omp parallel for if(bs > 100) // or small ks?
```
Likewise

t-vi commented:
Hm. Broadcasting. I only remembered bmm didn't... :/ Sorry, I'll get that fixed.

t-vi committed:
Fix broadcasting, guard openmp pragmas

t-vi committed:
disable bmm test for half

t-vi committed:
Merge branch 'master' into bmm_manysmall

t-vi commented:
So the two build failures are
- `FAIL: test_scalar_fusion (__main__.TestScript)` on `pr/pytorch-linux-trusty-py2.7`. This seems to happen to several, but not all builds on Jenkins. I don't know why, but could the test be flaky?
- On `pr/caffe2-py2-mkl-ubuntu16.04-test` there is a timeout for a distributed operation.

SsnL commented:
test_scalar_fusion is indeed flaky (see #11360)
The caffe2 test is also flaky :/ (see https://github.com/pytorch/pytorch/issues/8982)

soumith commented:
@t-vi it'd be good to also fold in changes from https://github.com/pytorch/pytorch/pull/11365 into this PR. cc: @mingfeima

soumith commented:
oh, I noticed that you already commented on that PR, my bad for not noticing,

t-vi committed:
Add MKL support

MKL support by Ma Mingfei, @mingfeima, errors by myself

ezyang commented:
The TH (not THC) implementation is dead now, right? Can you delete it?

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -222,6 +228,151 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+      auto s2 = s1[i];
+      for (int64_t j = 0; j < js; j++) {
+        scalar_t &r = r2[j];
+	r = 0;
```
nit: misdented

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -222,6 +228,151 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+      auto s2 = s1[i];
+      for (int64_t j = 0; j < js; j++) {
+        scalar_t &r = r2[j];
+	r *= beta;
```
nit: misdented

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -222,6 +228,151 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+  auto s0 = self.accessor<scalar_t, 3>();
+  auto m0 = mat2.accessor<scalar_t, 3>();
+
+#ifdef _OPENMP
```
It might be good to remark that better results may be had by calling MKL, to let a potential future optimizer who is not you try it out.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -222,6 +228,151 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+  } else { // split along batch dimension
+    if (is_bmm_out) {
+      for (int64_t b = 0; b < bs; b++) {
+	auto r = self_or_result.select(0, b);
```
nit: misdented

ezyang commented:
error is legit:

```
14:38:48 RuntimeError: Function '_th_bmm' starts with a single underscore and is configured to have a method on Tensor. Functions that start with  a single underscore should only be functions in the at:: namespace and not methods on Tensor!
```

Just delete `variant: method` from the cwrap

t-vi commented:
Thank you for your comments! I will improve the PR based on your feedback.
So I tried to adapt @mingfeima 's patch #11365 to have a `baddbmm__mkl` (`baddbmm_`) but it is difficult:
- The `_mkl`-suffix messes up the type translations (non-constness of self),
- @fritzo 's benchmark from the bug report seems to indicate that it is slow when I compile with mkl. :(
  I will try compare with the original #11365 PR to see whether that these are my adaptations or whether MKL is indeed slow in that case.

ezyang commented:
> The _mkl-suffix messes up the type translations (non-constness of self) should I just adapt the types or would a prefix be OK?,

Prefix is fine. The reason it's messing up is that it looks for `_out` or `_` as a suffix to decide what the types should be. So just don't put mkl at the very end and you'll be fine.

SsnL commented:
I'd just name it as `baddbmm_mkl_`

mingfeima commented:
@t-vi @soumith  fold #11365 would be better :)
sorry, i wasn't looking at `baddbmm` at the beginning, it's better to have it optimized with `bmm`.
i have been looking at [opennmt](https://github.com/OpenNMT/OpenNMT-py), `bmm` is used quite a lot in global attention computation.

from the optimization point of view, the optimization can be done in two manners.

1. mkl batched gemm: as #11365 
2. OpenMP + sequential gemm: do OpenMP parallelization across `batch_size` dimension and call sequential MKL gemm.

i actually did both. Performance gain from mkl batched gemm is only mediocre. `openmp` + `sequential gemm` actually performs much better than `mkl batched gemm` on unit benchmark. But once tested from PyTorch, it is not that good, needs further refinement with pytorch threading model here.

anyway, at this moment it is fair enough to stick `mkl batched gemm` for this PR. I checked implementations from other major frameworks, no better approach found.

@t-vi a few suggestions on this PR

1. use mkl to perform gemm. gemm is actually quite a complex topic, you have to deal with blocking, vectorization, etc.
2. i suppose you don't have to explicitly use `#pragma omp parallel for` now. Even if you need, use `parallel_for` wrapper from `/aten/src/ATen/Parallel.h` would be probably a better idea.
3. `aten/src/ATen/native/LinearAlgebra.cpp` is kind of like a unified entrance for all backends. So put mkl related functions under `aten/src/ATen/native/mkl/` would be better. And you can guard compilation with `AT_MKL_ENABLED()` and guard runtime selection with `at::hasMKL()`. Similarly, in case i figure out how to fix `openmp` + `sequential gemm` issue, it should go to `aten/src/ATen/native/cpu`.
 4. `baddbmm` and `bmm` can share the same underlying mkl kernels. `bmm` is a specialization of `baddbmm` when `alpha` = 1 and `beta` = 0.

t-vi commented:
@mingfeima  Thank you for the detailed comments!
I think I need some help more detail to follow your suggestions:
1. Use mkl to perform gemm means that I use your batched gemm? I have added that locally, but the performance seems bad compared to spelling out the matrix multiplication when I have very small matrices (as in fritzo's benchmark).
2. for non-mkl, I'm not sure how to avoid parallel for (though using Parallel.h is much better than trying to do this myself, thanks for pointing that out!).
3. my impression was that CPU is for specialized, accellerated (like AVX) code and "generic" CPU code is in native (e.g. Embedding, ...)
4. I think I have that locally. :)

I currently run this on my laptop (Thinkpad with Intel Core i7-5600U on Linux), I could move it to my GPU host (i5-7500).

t-vi committed:
review comments, drop TH baddbmm

t-vi committed:
Merge branch 'master' into bmm_manysmall

mingfeima commented:
@t-vi perhaps you will get different result on Xeon.
I test benchmarks on server CPUs, a.k.a. Xeon. And the current most highend model is skylake 8180 with 56 cores. Desktop CPU is actually not the major optimization target of Intel. I suggest you try Xeon, they are available on clouds, e.g. AWS, etc.

mingfeima commented:
@t-vi Xeon takes a very important role in our roadmap, we do benchmark on Xeon at most of the occasions.
And indeed, matrix size is related, in case the gemm is really small and `batch_size` is not large enough, a.k.a., the total `bmm` has little computation, calling sequential gemm would be technically faster. Anyway, in case you are dealing with `bmm` of very little computation, the absolute time used is minimal, even you suffer some performance loss with batched gemm, it can be tolerated.

t-vi committed:
add back CPU baddbmm for THNN, other review comments

Refined the switch between batch-parallel naive and
series of mm.
Bypass MKL based on same criterion based on very rough
benchmark.
Add comment about the optimization.

t-vi commented:
Thank you all for your input, I hope you find the changes make good use of it!

t-vi committed:
Tighten check for MKL applicability.

t-vi committed:
handle empty tensors

t-vi commented:
I'm not really sure about the failing CI test, but other than that, I think the PR should be ready for review.
